{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ### Multinomial NB Classifer to Predict Categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from sklearn.model_selection import learning_curve\n",
    "from sklearn.model_selection import train_test_split, cross_val_score, ShuffleSplit\n",
    "from sklearn.metrics import classification_report\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd, seaborn as sns, numpy as np, matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer, TfidfVectorizer \n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "\n",
    "from sklearn.pipeline import Pipeline\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "multinom_pipeline = Pipeline([\n",
    "    ('countvec', CountVectorizer(stop_words='english', ngram_range=(1,3))),\n",
    "    ('tfidfvec', TfidfTransformer()),\n",
    "    ('clfr', MultinomialNB())\n",
    "])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Setup a train/test/split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jovyan/ipynb'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import chdir\n",
    "chdir('/home/jovyan/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import library.db_helper as db\n",
    "import library.functions as fy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from sklearn.cross_validation import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2439, 1)"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = '''\n",
    "SELECT text\n",
    "FROM page;\n",
    "'''\n",
    "\n",
    "X = db.query_to_dataframe(query)\n",
    "X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>this article is an orphan as no other articles...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>this article has multiple issues please help i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>this article has multiple issues please help i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>a block diagram of the cmac system for a singl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>choose and book was an e booking software appl...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                text\n",
       "0  this article is an orphan as no other articles...\n",
       "1  this article has multiple issues please help i...\n",
       "2  this article has multiple issues please help i...\n",
       "3  a block diagram of the cmac system for a singl...\n",
       "4  choose and book was an e booking software appl..."
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = '''\n",
    "SELECT * \n",
    "FROM category_page;\n",
    "'''\n",
    "\n",
    "cat_page_df = db.query_to_dataframe(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2444"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat_page_df['page_pid'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "pca_article = fy.beautify_html_article('Principal_component_analysis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'pca of a multivariate gaussian distribution centered at   with a standard deviation of  in roughly the     direction and of  in the orthogonal direction the vectors shown are the eigenvectors of the covariance matrix scaled by the square root of the corresponding eigenvalue and shifted so their tails are at the mean principal component analysis pca is a statistical procedure that uses an orthogonal transformation to convert a set of observations of possibly correlated variables into a set of values of linearly uncorrelated variables called principal components or sometimes principal modes of variation the number of principal components is less than or equal to the smaller of the number of original variables or the number of observations this transformation is defined in such a way that the first principal component has the largest possible variance that is accounts for as much of the variability in the data as possible and each succeeding component in turn has the highest variance possible under the constraint that it is orthogonal to the preceding components the resulting vectors are an uncorrelated orthogonal basis set pca is sensitive to the relative scaling of the original variables pca was invented in  by karl pearson  as an analogue of the principal axis theorem in mechanics it was later independently developed and named by harold hotelling in the s  depending on the field of application it is also named the discrete kosambi karhunen lo√®ve transform klt in signal processing the hotelling transform in multivariate quality control proper orthogonal decomposition pod in mechanical engineering singular value decomposition svd of x golub and van loan  eigenvalue decomposition evd of xtx in linear algebra factor analysis for a discussion of the differences between pca and factor analysis see ch  of  eckart young theorem harman  or schmidt mirsky theorem in psychometrics empirical orthogonal functions eof in meteorological science empirical eigenfunction decomposition sirovich  empirical component analysis lorenz  quasiharmonic modes brooks et al  spectral decomposition in noise and vibration and empirical modal analysis in structural dynamics pca is mostly used as a tool in exploratory data analysis and for making predictive models it s often used to visualize genetic distance and relatedness between populations pca can be done by eigenvalue decomposition of a data covariance or correlation matrix or singular value decomposition of a data matrix usually after mean centering and normalizing or using z scores the data matrix for each attribute  the results of a pca are usually discussed in terms of component scores sometimes called factor scores the transformed variable values corresponding to a particular data point and loadings the weight by which each standardized original variable should be multiplied to get the component score  pca is the simplest of the true eigenvector based multivariate analyses often its operation can be thought of as revealing the internal structure of the data in a way that best explains the variance in the data if a multivariate dataset is visualised as a set of coordinates in a high dimensional data space  axis per variable pca can supply the user with a lower dimensional picture a projection of this object when viewed from its most informative viewpoint this is done by using only the first few principal components so that the dimensionality of the transformed data is reduced pca is closely related to factor analysis factor analysis typically incorporates more domain specific assumptions about the underlying structure and solves eigenvectors of a slightly different matrix pca is also related to canonical correlation analysis cca cca defines coordinate systems that optimally describe the cross covariance between two datasets while pca defines a new orthogonal coordinate system that optimally describes variance in a single dataset   contents intuition details  first component  further components  covariances  dimensionality reduction  singular value decomposition further considerations table of symbols and abbreviations properties and limitations of pca  properties    limitations  pca and information theory computing pca using the covariance method  organize the data set  calculate the empirical mean  calculate the deviations from the mean  find the covariance matrix  find the eigenvectors and eigenvalues of the covariance matrix  rearrange the eigenvectors and eigenvalues  compute the cumulative energy content for each eigenvector  select a subset of the eigenvectors as basis vectors  convert the source data to z scores optional   project the z scores of the data onto the new basis derivation of pca using the covariance method  iterative computation  the nipals method  online sequential estimation pca and qualitative variables applications  interest rate derivatives portfolios  neuroscience relation between pca and k means clustering relation between pca and factor analysis correspondence analysis generalizations  nonlinear generalizations  multilinear generalizations  higher order  robustness weighted pca  robust pca via decomposition in low rank and sparse matrices  sparse pca similar techniques  independent component analysis  network component analysis software source code see also references further reading external linksintuition edit pca can be thought of as fitting an n dimensional ellipsoid to the data where each axis of the ellipsoid represents a principal component if some axis of the ellipsoid is small then the variance along that axis is also small and by omitting that axis and its corresponding principal component from our representation of the dataset we lose only a commensurately small amount of information to find the axes of the ellipsoid we must first subtract the mean of each variable from the dataset to center the data around the origin then we compute the covariance matrix of the data and calculate the eigenvalues and corresponding eigenvectors of this covariance matrix then we must normalize each of the orthogonal eigenvectors to become unit vectors once this is done each of the mutually orthogonal unit eigenvectors can be interpreted as an axis of the ellipsoid fitted to the data the proportion of the variance that each eigenvector represents can be calculated by dividing the eigenvalue corresponding to that eigenvector by the sum of all eigenvalues this procedure is sensitive to the scaling of the data and there is no consensus as to how to best scale the data to obtain optimal results details edit pca is mathematically defined as an orthogonal linear transformation that transforms the data to a new coordinate system such that the greatest variance by some projection of the data comes to lie on the first coordinate called the first principal component the second greatest variance on the second coordinate and so on  consider a data matrix x with column wise zero empirical mean the sample mean of each column has been shifted to zero where each of the n rows represents a different repetition of the experiment and each of the p columns gives a particular kind of feature say the results from a particular sensor mathematically the transformation is defined by a set of p dimensional vectors of weights or loadings w k w wp k displaystyle mathbf w   k w   dots w  p   k that map each row vector x i displaystyle mathbf x   i of x to a new vector of principal component scores t i t tm i displaystyle mathbf t   i t   dots t  m   i given bytk i x i w k fori  nk  m displaystyle t  k   i mathbf x   i cdot mathbf w   k qquad mathrm for qquad i  dots n qquad k  dots m in such a way that the individual variables of t considered over the data set successively inherit the maximum possible variance from x with each loading vector w constrained to be a unit vector first component edit in order to maximize variance the first loading vector w  thus has to satisfyw  argmax w  i t i  argmax w  i x i w  displaystyle mathbf w    underset vert mathbf w vert  operatorname arg max left sum   i left t   right   i  right underset vert mathbf w vert  operatorname arg max left sum   i left mathbf x   i cdot mathbf w right  right equivalently writing this in matrix form givesw  argmax w  xw  argmax w  wtxtxw displaystyle mathbf w    underset vert mathbf w vert  operatorname arg max vert mathbf xw vert  underset vert mathbf w vert  operatorname arg max left mathbf w t mathbf x t mathbf xw right since w  has been defined to be a unit vector it equivalently also satisfiesw  argmax wtxtxwwtw displaystyle mathbf w    operatorname arg max left frac mathbf w t mathbf x t mathbf xw mathbf w t mathbf w right the quantity to be maximised can be recognised as a rayleigh quotient a standard result for a positive semidefinite matrix such as xtx is that the quotient s maximum possible value is the largest eigenvalue of the matrix which occurs when w is the corresponding eigenvector with w  found the first component of a data vector x i can then be given as a score t i x i w  in the transformed co ordinates or as the corresponding vector in the original variables x i w  w  further components edit the kth component can be found by subtracting the first k  principal components from x x k x s k xw s w s t displaystyle mathbf hat x   k mathbf x sum   s  k  mathbf x mathbf w   s mathbf w   s rm t and then finding the loading vector which extracts the maximum variance from this new data matrixw k argmax w  x kw  argmax wtx ktx kwwtw displaystyle mathbf w   k underset vert mathbf w vert  operatorname arg max left vert mathbf hat x   k mathbf w vert  right operatorname arg max left tfrac mathbf w t mathbf hat x   k t mathbf hat x   k mathbf w mathbf w t mathbf w right it turns out that this gives the remaining eigenvectors of xtx with the maximum values for the quantity in brackets given by their corresponding eigenvalues thus the loading vectors are eigenvectors of xtx the kth component of a data vector x i can therefore be given as a score tk i x i w k in the transformed co ordinates or as the corresponding vector in the space of the original variables x i w k w k where w k is the kth eigenvector of xtx the full principal components decomposition of x can therefore be given ast xw displaystyle mathbf t mathbf x mathbf w where w is a p by p matrix whose columns are the eigenvectors of xtx the transpose of w is sometimes called the whitening or sphering transformation covariances edit xtx itself can be recognised as proportional to the empirical sample covariance matrix of the dataset x the sample covariance q between two of the different principal components over the dataset is given by q pc j pc k xw j t xw k w j txtxw k w j t k w k  k w j tw k displaystyle begin aligned q mathrm pc   j mathrm pc   k propto mathbf x mathbf w   j t mathbf x mathbf w   k mathbf w   j t mathbf x t mathbf x mathbf w   k mathbf w   j t lambda   k mathbf w   k lambda   k mathbf w   j t mathbf w   k end aligned where the eigenvalue property of w k has been used to move from line  to line  however eigenvectors w j and w k corresponding to eigenvalues of a symmetric matrix are orthogonal if the eigenvalues are different or can be orthogonalised if the vectors happen to share an equal repeated value the product in the final line is therefore zero there is no sample covariance between different principal components over the dataset another way to characterise the principal components transformation is therefore as the transformation to coordinates which diagonalise the empirical sample covariance matrix in matrix form the empirical covariance matrix for the original variables can be writtenq xtx wŒªwt displaystyle mathbf q propto mathbf x t mathbf x mathbf w mathbf lambda mathbf w t the empirical covariance matrix between the principal components becomeswtqw wtwŒªwtw  displaystyle mathbf w t mathbf q mathbf w propto mathbf w t mathbf w mathbf lambda mathbf w t mathbf w mathbf lambda where  is the diagonal matrix of eigenvalues  k of xtx  k being equal to the sum of the squares over the dataset associated with each component k  k œÉi tk i œÉi x i w k  dimensionality reduction edit the transformation t x w maps a data vector x i from an original space of p variables to a new space of p variables which are uncorrelated over the dataset however not all the principal components need to be kept keeping only the first l principal components produced by using only the first l loading vectors gives the truncated transformationtl xwl displaystyle mathbf t   l mathbf x mathbf w   l where the matrix tl now has n rows but only l columns in other words pca learns a linear transformation t wtx x rp t rl displaystyle t w t x x in r p t in r l where the columns of p l matrix w form an orthogonal basis for the l features the components of representation t that are decorrelated  by construction of all the transformed data matrices with only l columns this score matrix maximises the variance in the original data that has been preserved while minimising the total squared reconstruction error twt tlwlt  displaystyle mathbf t mathbf w t mathbf t   l mathbf w   l t     or x xl  displaystyle mathbf x mathbf x   l     a principal components analysis scatterplot of y str haplotypes calculated from repeat count values for  y chromosomal str markers from  individuals pca has successfully found linear combinations of the different markers that separate out different clusters corresponding to different lines of individuals y chromosomal genetic descent such dimensionality reduction can be a very useful step for visualising and processing high dimensional datasets while still retaining as much of the variance in the dataset as possible for example selecting l  and keeping only the first two principal components finds the two dimensional plane through the high dimensional dataset in which the data is most spread out so if the data contains clusters these too may be most spread out and therefore most visible to be plotted out in a two dimensional diagram whereas if two directions through the data or two of the original variables are chosen at random the clusters may be much less spread apart from each other and may in fact be much more likely to substantially overlay each other making them indistinguishable similarly in regression analysis the larger the number of explanatory variables allowed the greater is the chance of overfitting the model producing conclusions that fail to generalise to other datasets one approach especially when there are strong correlations between different possible explanatory variables is to reduce them to a few principal components and then run the regression against them a method called principal component regression dimensionality reduction may also be appropriate when the variables in a dataset are noisy if each column of the dataset contains independent identically distributed gaussian noise then the columns of t will also contain similarly identically distributed gaussian noise such a distribution is invariant under the effects of the matrix w which can be thought of as a high dimensional rotation of the co ordinate axes however with more of the total variance concentrated in the first few principal components compared to the same noise variance the proportionate effect of the noise is less the first few components achieve a higher signal to noise ratio pca thus can have the effect of concentrating much of the signal into the first few principal components which can usefully be captured by dimensionality reduction while the later principal components may be dominated by noise and so disposed of without great loss singular value decomposition edit the principal components transformation can also be associated with another matrix factorization the singular value decomposition svd of x x uœÉwt displaystyle mathbf x mathbf u mathbf sigma mathbf w t here  is an n by p rectangular diagonal matrix of positive numbers  k called the singular values of x u is an n by n matrix the columns of which are orthogonal unit vectors of length n called the left singular vectors of x and w is a p by p whose columns are orthogonal unit vectors of length p and called the right singular vectors of x in terms of this factorization the matrix xtx can be writtenxtx wœÉutuœÉwt wœÇwt displaystyle begin aligned mathbf x t mathbf x mathbf w mathbf sigma mathbf u t mathbf u mathbf sigma mathbf w t mathbf w mathbf sigma  mathbf w t end aligned comparison with the eigenvector factorization of xtx establishes that the right singular vectors w of x are equivalent to the eigenvectors of xtx while the singular values  k of  are equal to the square roots of the eigenvalues  k of xtx using the singular value decomposition the score matrix t can be writtent xw uœÉwtw u displaystyle begin aligned mathbf t mathbf x mathbf w mathbf u mathbf sigma mathbf w t mathbf w mathbf u mathbf sigma end aligned so each column of t is given by one of the left singular vectors of x multiplied by the corresponding singular value this form is also the polar decomposition of t efficient algorithms exist to calculate the svd of x without having to form the matrix xtx so computing the svd is now the standard way to calculate a principal components analysis from a data matrix citation needed unless only a handful of components are required as with the eigen decomposition a truncated n l score matrix tl can be obtained by considering only the first l largest singular values and their singular vectors tl ulœÉl xwl displaystyle mathbf t   l mathbf u   l mathbf sigma   l mathbf x mathbf w   l the truncation of a matrix m or t using a truncated singular value decomposition in this way produces a truncated matrix that is the nearest possible matrix of rank l to the original matrix in the sense of the difference between the two having the smallest possible frobenius norm a result known as the eckart young theorem  further considerations edit given a set of points in euclidean space the first principal component corresponds to a line that passes through the multidimensional mean and minimizes the sum of squares of the distances of the points from the line the second principal component corresponds to the same concept after all correlation with the first principal component has been subtracted from the points the singular values in  are the square roots of the eigenvalues of the matrix xtx each eigenvalue is proportional to the portion of the variance more correctly of the sum of the squared distances of the points from their multidimensional mean that is correlated with each eigenvector the sum of all the eigenvalues is equal to the sum of the squared distances of the points from their multidimensional mean pca essentially rotates the set of points around their mean in order to align with the principal components this moves as much of the variance as possible using an orthogonal transformation into the first few dimensions the values in the remaining dimensions therefore tend to be small and may be dropped with minimal loss of information see below pca is often used in this manner for dimensionality reduction pca has the distinction of being the optimal orthogonal transformation for keeping the subspace that has largest variance as defined above this advantage however comes at the price of greater computational requirements if compared for example and when applicable to the discrete cosine transform and in particular to the dct ii which is simply known as the dct nonlinear dimensionality reduction techniques tend to be more computationally demanding than pca pca is sensitive to the scaling of the variables if we have just two variables and they have the same sample variance and are positively correlated then the pca will entail a rotation by  and the loadings for the two variables with respect to the principal component will be equal but if we multiply all values of the first variable by  then the first principal component will be almost the same as that variable with a small contribution from the other variable whereas the second component will be almost aligned with the second original variable this means that whenever the different variables have different units like temperature and mass pca is a somewhat arbitrary method of analysis different results would be obtained if one used fahrenheit rather than celsius for example note that pearson s original paper was entitled on lines and planes of closest fit to systems of points in space in space implies physical euclidean space where such concerns do not arise one way of making the pca less arbitrary is to use variables scaled so as to have unit variance by standardizing the data and hence use the autocorrelation matrix instead of the autocovariance matrix as a basis for pca however this compresses or expands the fluctuations in all dimensions of the signal space to unit variance mean subtraction a k a mean centering is necessary for performing pca to ensure that the first principal component describes the direction of maximum variance if mean subtraction is not performed the first principal component might instead correspond more or less to the mean of the data a mean of zero is needed for finding a basis that minimizes the mean square error of the approximation of the data  mean centering is unnecessary if performing a principal components analysis on a correlation matrix as the data are already centered after calculating correlations correlations are derived from the cross product of two standard scores z scores or statistical moments hence the name pearson product moment correlation also see the article by kromrey foster johnson  on mean centering in moderated regression much ado about nothing an autoencoder neural network with a linear hidden layer is similar to pca upon convergence the weight vectors of the k neurons in the hidden layer will form a basis for the space spanned by the first k principal components unlike pca this technique will not necessarily produce orthogonal vectors pca is a popular primary technique in pattern recognition it is not however optimized for class separability  however it has been used to quantify the distance between two or more classes by calculating center of mass for each class in principal component space and reporting euclidean distance between center of mass of two or more classes  the linear discriminant analysis is an alternative which is optimized for class separability table of symbols and abbreviations edit symbolmeaningdimensionsindicesx xij displaystyle mathbf x x  ij data matrix consisting of the set of all data vectors one vector per rown p displaystyle n times p i  n displaystyle i  ldots n j  p displaystyle j  ldots p n displaystyle n the number of row vectors in the data set  displaystyle  times  scalarp displaystyle p the number of elements in each row vector dimension   displaystyle  times  scalarl displaystyle l the number of dimensions in the dimensionally reduced subspace  l p displaystyle  leq l leq p   displaystyle  times  scalaru uj displaystyle mathbf u u  j vector of empirical means one mean for each column j of the data matrixp  displaystyle p times  j  p displaystyle j  ldots p s sj displaystyle mathbf s s  j vector of empirical standard deviations one standard deviation for each column j of the data matrixp  displaystyle p times  j  p displaystyle j  ldots p h hi displaystyle mathbf h h  i vector of all  s n displaystyle  times n i  n displaystyle i  ldots n b bij displaystyle mathbf b b  ij deviations from the mean of each column j of the data matrixn p displaystyle n times p i  n displaystyle i  ldots n j  p displaystyle j  ldots p z zij displaystyle mathbf z z  ij z scores computed using the mean and standard deviation for each row m of the data matrixn p displaystyle n times p i  n displaystyle i  ldots n j  p displaystyle j  ldots p c cjj displaystyle mathbf c c  jj covariance matrixp p displaystyle p times p j  p displaystyle j  ldots p j  p displaystyle j  ldots p r rjj displaystyle mathbf r r  jj correlation matrixp p displaystyle p times p j  p displaystyle j  ldots p j  p displaystyle j  ldots p v vjj displaystyle mathbf v v  jj matrix consisting of the set of all eigenvectors of c one eigenvector per columnp p displaystyle p times p j  p displaystyle j  ldots p j  p displaystyle j  ldots p d djj displaystyle mathbf d d  jj diagonal matrix consisting of the set of all eigenvalues of c along its principal diagonal and  for all other elementsp p displaystyle p times p j  p displaystyle j  ldots p j  p displaystyle j  ldots p w wjl displaystyle mathbf w w  jl matrix of basis vectors one vector per column where each basis vector is one of the eigenvectors of c and where the vectors in w are a sub set of those in vp l displaystyle p times l j  p displaystyle j  ldots p l  l displaystyle l  ldots l t til displaystyle mathbf t t  il matrix consisting of n row vectors where each vector is the projection of the corresponding data vector from matrix x onto the basis vectors contained in the columns of matrix w n l displaystyle n times l i  n displaystyle i  ldots n l  l displaystyle l  ldots l properties and limitations of pca edit properties  edit property  for any integer q  q p consider the orthogonal linear transformationy b x displaystyle y mathbf b x where y displaystyle y is a q element vector and b displaystyle mathbf b is a q p matrix and let œÉy b œÉb displaystyle mathbf sigma   y mathbf b mathbf sigma mathbf b be the variance covariance matrix for y displaystyle y then the trace of œÉy displaystyle mathbf sigma   y denoted tr œÉy displaystyle text tr mathbf sigma   y is maximized by taking b aq displaystyle mathbf b mathbf a   q where aq displaystyle mathbf a   q consists of the first q columns of a displaystyle mathbf a b displaystyle mathbf b is the transposition of b displaystyle mathbf b property  consider again the orthonormal transformationy b x displaystyle y mathbf b x with x b a displaystyle x mathbf b mathbf a and œÉy displaystyle mathbf sigma   y defined as before then tr œÉy displaystyle text tr mathbf sigma   y is minimized by taking b aq displaystyle mathbf b mathbf a   q where aq displaystyle mathbf a   q consists of the last q columns of a displaystyle mathbf a the statistical implication of this property is that the last few pcs are not simply unstructured left overs after removing the important pcs because these last pcs have variances as small as possible they are useful in their own right they can help to detect unsuspected near constant linear relationships between the elements of x and they may also be useful in regression in selecting a subset of variables from x and in outlier detection property  spectral decomposition of   ŒªŒ±Œ± ŒªpŒ±pŒ±p displaystyle mathbf sigma lambda    alpha    alpha    cdots lambda   p alpha   p alpha   p before we look at its usage we first look at diagonal elements var xj k pŒªkŒ±kj displaystyle text var x  j sum   k  p lambda   k alpha   kj  then perhaps the main statistical implication of the result is that not only can we decompose the combined variances of all the elements of x into decreasing contributions due to each pc but we can also decompose the whole covariance matrix into contributions ŒªkŒ±kŒ±k displaystyle lambda   k alpha   k alpha   k from each pc although not strictly decreasing the elements of ŒªkŒ±kŒ±k displaystyle lambda   k alpha   k alpha   k will tend to become smaller as k displaystyle k increases as ŒªkŒ±kŒ±k displaystyle lambda   k alpha   k alpha   k is nonincreasing for increasing k displaystyle k whereas the elements of Œ±k displaystyle alpha   k tend to stay about the same size because of the normalization constraints Œ±k Œ±k  k  p displaystyle alpha   k alpha   k  k  cdots p limitations edit as noted above the results of pca depend on the scaling of the variables a scale invariant form of pca has been developed  the applicability of pca is limited by certain assumptions  made in its derivation pca and information theory edit dimensionality reduction loses information in general pca based dimensionality reduction tends to minimize that information loss under certain signal and noise models under the assumption thatx s n displaystyle mathbf x mathbf s mathbf n i e that the data vector x displaystyle mathbf x is the sum of the desired information bearing signal s displaystyle mathbf s and a noise signal n displaystyle mathbf n one can show that pca can be optimal for dimensionality reduction from an information theoretic point of view in particular linsker showed that if s displaystyle mathbf s is gaussian and n displaystyle mathbf n is gaussian noise with a covariance matrix proportional to the identity matrix the pca maximizes the mutual information i y s displaystyle i mathbf y mathbf s between the desired information s displaystyle mathbf s and the dimensionality reduced output y wltx displaystyle mathbf y mathbf w   l t mathbf x  if the noise is still gaussian and has a covariance matrix proportional to the identity matrix i e the components of the vector n displaystyle mathbf n are iid but the information bearing signal s displaystyle mathbf s is non gaussian which is a common scenario pca at least minimizes an upper bound on the information loss which is defined as   i x s i y s displaystyle i mathbf x mathbf s i mathbf y mathbf s the optimality of pca is also preserved if the noise n displaystyle mathbf n is iid and at least more gaussian in terms of the kullback leibler divergence than the information bearing signal s displaystyle mathbf s  in general even if the above signal model holds pca loses its information theoretic optimality as soon as the noise n displaystyle mathbf n becomes dependent computing pca using the covariance method edit the following is a detailed description of pca using the covariance method see also here as opposed to the correlation method  but note that it is better to use the singular value decomposition using standard software citation needed the goal is to transform a given data set x of dimension p to an alternative data set y of smaller dimension l equivalently we are seeking to find the matrix y where y is the kosambi karhunen lo√®ve transform klt of matrix x y klt x displaystyle mathbf y mathbb klt mathbf x organize the data set edit suppose you have data comprising a set of observations of p variables and you want to reduce the data so that each observation can be described with only l variables l p suppose further that the data are arranged as a set of n data vectors x xn displaystyle mathbf x    ldots mathbf x   n with each xi displaystyle mathbf x   i representing a single grouped observation of the p variables write x xn displaystyle mathbf x    ldots mathbf x   n as row vectors each of which has p columns place the row vectors into a single matrix x of dimensions n p calculate the empirical mean edit find the empirical mean along each column j  p place the calculated mean values into an empirical mean vector u of dimensions p  uj n i nxij displaystyle u  j  over n sum   i  n x  ij calculate the deviations from the mean edit mean subtraction is an integral part of the solution towards finding a principal component basis that minimizes the mean square error of approximating the data  hence we proceed by centering the data as follows subtract the empirical mean vector u from each row of the data matrix x store mean subtracted data in the n p matrix b b x hut displaystyle mathbf b mathbf x mathbf h mathbf u t where h is an n  column vector of all s hi for i  n displaystyle h  i  qquad qquad text for i  ldots n find the covariance matrix edit find the p p empirical covariance matrix c from the outer product of matrix b with itself c n b b displaystyle mathbf c  over n  mathbf b otimes mathbf b where displaystyle is the conjugate transpose operator note that if b consists entirely of real numbers which is the case in many applications the conjugate transpose is the same as the regular transpose please note that outer products apply to vectors for tensor cases we should apply tensor products but the covariance matrix in pca is a sum of outer products between its sample vectors indeed it could be represented as b b see the covariance matrix sections on the discussion page for more information the reasoning behind using n  instead of n to calculate the covariance is bessel s correctionfind the eigenvectors and eigenvalues of the covariance matrix edit compute the matrix v of eigenvectors which diagonalizes the covariance matrix c v cv d displaystyle mathbf v  mathbf c mathbf v mathbf d where d is the diagonal matrix of eigenvalues of c this step will typically involve the use of a computer based algorithm for computing eigenvectors and eigenvalues these algorithms are readily available as sub components of most matrix algebra systems such as sas  r matlab   mathematica  scipy idl interactive data language or gnu octave as well as opencv matrix d will take the form of an p p diagonal matrix wheredkl Œªkfor k l displaystyle d  kl lambda   k qquad text for k l is the jth eigenvalue of the covariance matrix c anddkl for k l displaystyle d  kl  qquad text for k neq l matrix v also of dimension p p contains p column vectors each of length p which represent the p eigenvectors of the covariance matrix c the eigenvalues and eigenvectors are ordered and paired the jth eigenvalue corresponds to the jth eigenvector matrix v denotes the matrix of right eigenvectors as opposed to left eigenvectors in general the matrix of right eigenvectors need not be the conjugate transpose of the matrix of left eigenvectors rearrange the eigenvectors and eigenvalues edit sort the columns of the eigenvector matrix v and eigenvalue matrix d in order of decreasing eigenvalue make sure to maintain the correct pairings between the columns in each matrix compute the cumulative energy content for each eigenvector edit the eigenvalues represent the distribution of the source data s energy clarification needed among each of the eigenvectors where the eigenvectors form a basis for the data the cumulative energy content g for the jth eigenvector is the sum of the energy content across all of the eigenvalues from  through j gj k jdkkforj  p displaystyle g  j sum   k  j d  kk qquad mathrm for qquad j  dots p citation needed select a subset of the eigenvectors as basis vectors edit save the first l columns of v as the p l matrix w wkl vklfork  pl  l displaystyle w  kl v  kl qquad mathrm for qquad k  dots p qquad l  dots l where l p displaystyle  leq l leq p use the vector g as a guide in choosing an appropriate value for l the goal is to choose a value of l as small as possible while achieving a reasonably high value of g on a percentage basis for example you may want to choose l so that the cumulative energy g is above a certain threshold like  percent in this case choose the smallest value of l such thatglgp   displaystyle frac g  l g  p geq   convert the source data to z scores optional edit create an p  empirical standard deviation vector s from the square root of each element along the main diagonal of the diagonalized covariance matrix c note that scaling operations do not commute with the klt thus we must scale by the variances of the already decorrelated vector which is the diagonal of c s sj cjj for j  p displaystyle mathbf s s  j left sqrt c  jj right qquad text for j  ldots p calculate the n p z score matrix z bh st displaystyle mathbf z mathbf b over mathbf h cdot mathbf s t divide element by element note while this step is useful for various applications as it normalizes the data set with respect to its variance it is not integral part of pca kltproject the z scores of the data onto the new basis edit the projected vectors are the columns of the matrixt z w klt x displaystyle mathbf t mathbf z cdot mathbf w mathbb klt mathbf x the rows of matrix t represent the kosambi karhunen lo√®ve transforms klt of the data vectors in the rows of matrix x derivation of pca using the covariance method edit let x be a d dimensional random vector expressed as column vector without loss of generality assume x has zero mean we want to find displaystyle ast a d d orthonormal transformation matrix p so that px has a diagonal covariance matrix i e px is a random vector with all its distinct components pairwise uncorrelated a quick computation assuming p displaystyle p were unitary yields cov px e px px e px x p p e xx p p cov x p  displaystyle begin aligned operatorname cov px mathbb e px px mathbb e px x p p mathbb e xx p p operatorname cov x p  end aligned hence displaystyle ast holds if and only if cov x displaystyle operatorname cov x were diagonalisable by p displaystyle p this is very constructive as cov x is guaranteed to be a non negative definite matrix and thus is guaranteed to be diagonalisable by some unitary matrix iterative computation edit in practical implementations especially with high dimensional data large p the covariance method is rarely used because it is not efficient one way to compute the first principal component efficiently  is shown in the following pseudo code for a data matrix x with zero mean without ever computing its covariance matrix r a random vector of length pdo c times s  a vector of length p for each row x x displaystyle mathbf x in mathbf x s s x r x displaystyle mathbf s mathbf s mathbf x cdot mathbf r mathbf x r s s displaystyle mathbf r frac mathbf s mathbf s return rthis algorithm is simply an efficient way of calculating xtx r normalizing and placing the result back in r power iteration it avoids the np operations of calculating the covariance matrix r will typically get close to the first principal component of x within a small number of iterations c the magnitude of s will be larger after each iteration convergence can be detected when it increases by an amount too small for the precision of the machine subsequent principal components can be computed by subtracting component r from x see gram schmidt and then repeating this algorithm to find the next principal component however this simple approach is not numerically stable if more than a small number of principal components are required because imprecisions in the calculations will additively affect the estimates of subsequent principal components more advanced methods build on this basic idea as with the closely related lanczos algorithm one way to compute the eigenvalue that corresponds with each principal component is to measure the difference in mean squared distance between the rows and the centroid before and after subtracting out the principal component the eigenvalue that corresponds with the component that was removed is equal to this difference the nipals method edit non linear iterative partial least squares nipals is an algorithm for computing the first few components in a principal component or partial least squares analysis for very high dimensional datasets such as those generated in the omics sciences e g genomics metabolomics it is usually only necessary to compute the first few pcs the non linear iterative partial least squares nipals algorithm calculates t and wt from x the outer product twt can then be subtracted from x leaving the residual matrix e this can be then used to calculate subsequent pcs  this results in a dramatic reduction in computational time since calculation of the covariance matrix is avoided however for large data matrices or matrices that have a high degree of column collinearity nipals suffers from loss of orthogonality due to machine precision limitations accumulated in each iteration step  a gram schmidt gs re orthogonalization algorithm is applied to both the scores and the loadings at each iteration step to eliminate this loss of orthogonality  online sequential estimation edit in an online or streaming situation with data arriving piece by piece rather than being stored in a single batch it is useful to make an estimate of the pca projection that can be updated sequentially this can be done efficiently but requires different algorithms  pca and qualitative variables edit in pca it is common that we want to introduce qualitative variables as supplementary elements for example many quantitative variables have been measured on plants for these plants some qualitative variables are available as for example the species to which the plant belongs these data were subjected to pca for quantitative variables when analyzing the results it is natural to connect the principal components to the qualitative variable species for this the following results are produced identification on the factorial planes of the different species e g using different colors representation on the factorial planes of the centers of gravity of plants belonging to the same species for each center of gravity and each axis p value to judge the significance of the difference between the center of gravity and origin these results are what is called introducing a qualitative variable as supplementary element this procedure is detailed in and husson l pag√®s  and pag√®s  few software offer this option in an automatic way this is the case of spad that historically following the work of ludovic lebart was the first to propose this option and the r package factominer applications edit interest rate derivatives portfolios edit principal component analysis can be directly applied to the risk management of interest rate derivatives portfolios  trading multiple swap instruments which are usually a function of   other market quotable swap instruments is sought to be reduced to usually  or  principal components representing the path of interest rates on a macro basis converting risks to be represented as those to factor loadings or multipliers provides assessments and understanding beyond that available to simply collectively viewing risks to individual   buckets neuroscience edit a variant of principal components analysis is used in neuroscience to identify the specific properties of a stimulus that increase a neuron s probability of generating an action potential  this technique is known as spike triggered covariance analysis in a typical application an experimenter presents a white noise process as a stimulus usually either as a sensory input to a test subject or as a current injected directly into the neuron and records a train of action potentials or spikes produced by the neuron as a result presumably certain features of the stimulus make the neuron more likely to spike in order to extract these features the experimenter calculates the covariance matrix of the spike triggered ensemble the set of all stimuli defined and discretized over a finite time window typically on the order of  ms that immediately preceded a spike the eigenvectors of the difference between the spike triggered covariance matrix and the covariance matrix of the prior stimulus ensemble the set of all stimuli defined over the same length time window then indicate the directions in the space of stimuli along which the variance of the spike triggered ensemble differed the most from that of the prior stimulus ensemble specifically the eigenvectors with the largest positive eigenvalues correspond to the directions along which the variance of the spike triggered ensemble showed the largest positive change compared to the variance of the prior since these were the directions in which varying the stimulus led to a spike they are often good approximations of the sought after relevant stimulus features in neuroscience pca is also used to discern the identity of a neuron from the shape of its action potential spike sorting is an important procedure because extracellular recording techniques often pick up signals from more than one neuron in spike sorting one first uses pca to reduce the dimensionality of the space of action potential waveforms and then performs clustering analysis to associate specific action potentials with individual neurons pca as a dimension reduction technique is particularly suited to detect coordinated activities of large neuronal ensembles it has been used in determining collective variables i e order parameters during phase transitions in the brain  relation between pca and k means clustering edit it was asserted in   that the relaxed solution of k means clustering specified by the cluster indicators is given by the principal components and the pca subspace spanned by the principal directions is identical to the cluster centroid subspace however that pca is a useful relaxation of k means clustering was not a new result see for example  and it is straightforward to uncover counterexamples to the statement that the cluster centroid subspace is spanned by the principal directions  relation between pca and factor analysis edit principal component analysis creates variables that are linear combinations of the original variables the new variables have the property that the variables are all orthogonal the principal components can be used to find clusters in a set of data pca is a variance focused approach seeking to reproduce the total variable variance in which components reflect both common and unique variance of the variable pca is generally preferred for purposes of data reduction i e translating variable space into optimal factor space but not when the goal is to detect the latent construct or factors factor analysis is similar to principal component analysis  in that factor analysis also involves linear combinations of variables different from pca factor analysis is a correlation focused approach seeking to reproduce the inter correlations among variables in which the factors represent the common variance of variables excluding unique variance  in terms of the correlation matrix this corresponds with focusing on explaining the off diagonal terms i e shared co variance while pca focuses on explaining the terms that sit on the diagonal however as a side result when trying to reproduce the on diagonal terms pca also tends to fit relatively well the off diagonal correlations  results given by pca and factor analysis are very similar in most situations but this is not always the case and there are some problems where the results are significantly different factor analysis is generally used when the research purpose is detecting data structure i e latent constructs or factors or causal modeling correspondence analysis edit correspondence analysis ca was developed by jean paul benz√©cri  and is conceptually similar to pca but scales the data which should be non negative so that rows and columns are treated equivalently it is traditionally applied to contingency tables ca decomposes the chi squared statistic associated to this table into orthogonal factors  because ca is a descriptive technique it can be applied to tables for which the chi squared statistic is appropriate or not several variants of ca are available including detrended correspondence analysis and canonical correspondence analysis one special extension is multiple correspondence analysis which may be seen as the counterpart of principal component analysis for categorical data  generalizations edit nonlinear generalizations edit linear pca versus nonlinear principal manifolds  for visualization of breast cancer microarray data a configuration of nodes and d principal surface in the d pca linear manifold the dataset is curved and cannot be mapped adequately on a d principal plane b the distribution in the internal d non linear principal surface coordinates elmapd together with an estimation of the density of points c the same as b but for the linear d pca manifold pcad the basal breast cancer subtype is visualized more adequately with elmapd and some features of the distribution become better resolved in comparison to pcad principal manifolds are produced by the elastic maps algorithm data are available for public competition  software is available for free non commercial use  most of the modern methods for nonlinear dimensionality reduction find their theoretical and algorithmic roots in pca or k means pearson s original idea was to take a straight line or plane which will be the best fit to a set of data points principal curves and manifolds  give the natural geometric framework for pca generalization and extend the geometric interpretation of pca by explicitly constructing an embedded manifold for data approximation and by encoding using standard geometric projection onto the manifold as it is illustrated by fig see also the elastic map algorithm and principal geodesic analysis another popular generalization is kernel pca which corresponds to pca performed in a reproducing kernel hilbert space associated with a positive definite kernel multilinear generalizations edit in multilinear subspace learning  pca is generalized to multilinear pca mpca that extracts features directly from tensor representations mpca is solved by performing pca in each mode of the tensor iteratively mpca has been applied to face recognition gait recognition etc mpca is further extended to uncorrelated mpca non negative mpca and robust mpca higher order edit n way principal component analysis may be performed with models such as tucker decomposition parafac multiple factor analysis co inertia analysis statis and distatis robustness weighted pca edit while pca finds the mathematically optimal method as in minimizing the squared error it is sensitive to outliers in the data that produce large errors pca tries to avoid it therefore is common practice to remove outliers before computing pca however in some contexts outliers can be difficult to identify for example in data mining algorithms like correlation clustering the assignment of points to clusters and outliers is not known beforehand a recently proposed generalization of pca  based on a weighted pca increases robustness by assigning different weights to data objects based on their estimated relevancy robust pca via decomposition in low rank and sparse matrices edit robust principal component analysis rpca is a modification of the widely used statistical procedure principal component analysis pca which works well with respect to grossly corrupted observations sparse pca edit a particular disadvantage of pca is that the principal components are usually linear combinations of all input variables sparse pca overcomes this disadvantage by finding linear combinations that contain just a few input variables similar techniques edit independent component analysis edit independent component analysis ica is directed to similar problems as principal component analysis but finds additively separable components rather than successive approximations network component analysis edit given a matrix e displaystyle e it tries to decompose it into two matrices such that e ap displaystyle e ap a key difference from techniques such as pca and ica is that some of the entries of a displaystyle a are constrained to be  here p displaystyle p is termed the regulatory layer while in general such a decomposition can have multiple solutions they prove that if the following conditions are satisfied a displaystyle a has full column rankeach column of a displaystyle a must have at least l  displaystyle l  zeroes where l displaystyle l is the number of columns of a displaystyle a or alternatively the number of rows of p displaystyle p the justification for this criterion is that if a node is removed from the regulatory layer along with all the output nodes connected to it the result must still be characterized by a connectivity matrix with full column rank p displaystyle p must have full row rank then the decomposition is unique up to multiplication by a scalar  software source code edit analytica the built in eigendecomp function computes principal components datamelt a java free program that implements several classes to build pca analysis and to calculate eccentricity of random distributions elki includes pca for projection including robust variants of pca as well as pca based clustering algorithms igor pro the built in pca operation performs principal component analysis julia supports pca with the pca function in the multivariatestats packageknime a java based nodal arrenging software for analysis in this the nodes called pca pca compute pca apply pca inverse make it easily mathematica implements principal component analysis with the principalcomponents command using both covariance and correlation methods matlab statistics toolbox the functions princomp and pca rb give the principal components while the function pcares gives the residuals and reconstructed matrix for a low rank pca approximation matplotlib python library have a pca package in the mlab module mlpack provides an implementation of principal component analysis in c nag library principal components analysis is implemented via the gaa routine available in both the fortran versions of the library nmath proprietary numerical library containing pca for the net framework gnu octave free software computational environment mostly compatible with matlab the function princomp gives the principal component opencvoracle database c implemented via dbms data mining svds scoring mode by specifying setting value svds scoring pcaorange software integrates pca in its visual programming environment pca displays a scree plot degree of explained variance where user can interactively select the number of principal components origin contains pca in its pro version padasip python library that contains pca in its data preprocessing module partek genomics suite statistical software able to perform pca qlucore commercial software for analyzing multivariate data with instant response using pca r free statistical package the functions princomp and prcomp can be used for principal component analysis prcomp uses singular value decomposition which generally gives better numerical accuracy some packages that implement pca in r include but are not limited to ade vegan exposition and factominer scikit learn python library for machine learning which contains pca probabilistic pca kernel pca sparse pca and other techniques in the decomposition module weka java library for machine learning which contains modules for computing principal components see also edit correspondence analysis for contingency tables multiple correspondence analysis for qualitative variables factor analysis of mixed data for quantitative and qualitative variables canonical correlationcur matrix approximation can replace of low rank svd approximation detrended correspondence analysisdynamic mode decompositioneigenfaceexploratory factor analysis wikiversity factorial codefunctional principal component analysisgeometric data analysisindependent component analysiskernel pcalow rank approximationmatrix decompositionnon negative matrix factorizationnonlinear dimensionality reductionoja s rulepoint distribution model pca applied to morphometry and computer vision principal component analysis wikibooks principal component regressionsingular spectrum analysissingular value decompositionsparse pcatransform codingweighted least squaresreferences edit pearson k  on lines and planes of closest fit to systems of points in space pdf philosophical magazine     doi    hotelling h  analysis of a complex of statistical variables into principal components journal of educational psychology    and   hotelling h  relations between two sets of variates biometrika    a b jolliffe i t principal component analysis series springer series in statistics nd ed springer ny  xxix  p  illus isbn      abdi h williams l j  principal component analysis wiley interdisciplinary reviews computational statistics     doi   wics  shaw p j a  multivariate statistics for the environmental sciences hodder arnold isbn     page needed barnett t p r preisendorfer  origins and levels of monthly and seasonal forecast skill for united states surface air temperatures determined by canonical correlation analysis monthly weather review  hsu daniel sham m kakade and tong zhang  a spectral algorithm for learning hidden markov models arxiv   cs maint multiple names authors list link bengio y et al  representation learning a review and new perspectives pdf pattern analysis and machine intelligence     doi   tpami   a a miranda y a le borgne and g bontempi new routes from minimal approximation error to principal components volume  number  june  neural processing letters springer fukunaga keinosuke  introduction to statistical pattern recognition elsevier isbn     alizadeh elaheh lyons samanthe m castle jordan m prasad ashok  measuring systematic changes in invasive cancer cell shape using zernike moments integrative biology     doi   ciba jolliffe i t  principal component analysis second edition springer verlag isbn      leznik m tofallis c  uhra herts ac uk bitstream handle   s pdf estimating invariant principal components using diagonal regression jonathon shlens a tutorial on principal component analysis linsker ralph march  self organization in a perceptual network ieee computer     doi     deco obradovic  an information theoretic approach to neural computing new york ny springer plumbley mark  information theory and unsupervised neural networks tech note geiger bernhard kubin gernot january  signal enhancement as minimization of relevant information loss proc itg conf on systems communication and coding arxiv   engineering statistics handbook section     retrieved  january  a a miranda y a le borgne and g bontempi new routes from minimal approximation error to principal components volume  number  june  neural processing letters springer http support sas com documentation cdl en statug  html default viewer htm statug princomp sect htm eig function matlab documentation matlab pca based face recognition software eigenvalues function mathematica documentation roweis sam em algorithms for pca and spca advances in neural information processing systems ed michael i jordan michael j kearns and sara a solla the mit press  geladi paul kowalski bruce  partial least squares regression a tutorial analytica chimica acta    doi        kramer r  chemometric techniques for quantitative analysis new york crc press andrecut m  parallel gpu implementation of iterative pca algorithms journal of computational biology     pmid  doi   cmb   warmuth m k kuzmin d  randomized online pca algorithms with regret bounds that are logarithmic in the dimension journal of machine learning research    the pricing and hedging of interest rate derivatives a practical guide to swaps j h m darbyshire  isbn   brenner n bialek w de ruyter van steveninck r r  jirsa victor friedrich r haken herman kelso scott  a theoretical model of phase transitions in the human brain biological cybernetics     pmid  doi   bf h zha c ding m gu x he and h d simon dec  spectral relaxation for k means clustering pdf neural information processing systems vol  nips  vancouver canada   cs maint uses authors parameter link chris ding and xiaofeng he july  k means clustering via principal component analysis pdf proc of int l conf machine learning icml    cs maint uses authors parameter link drineas p a frieze r kannan s vempala v vinay  clustering large graphs via the singular value decomposition pdf machine learning    doi   b mach    retrieved    cohen m s elder c musco c musco m persu  dimensionality reduction for k means clustering and low rank approximation appendix b arxiv   ijsmi editor    tutorial factor analysis revisited an overview with the help of spss sas and r packages international journal of statistics and medical informatics   timothy a brown confirmatory factor analysis for applied research methodology in the social sciences guilford press  i t jolliffe principal component analysis second edition chapter   benz√©cri j p  l analyse des donn√©es volume ii l analyse des correspondances paris france dunod greenacre michael  theory and applications of correspondence analysis london academic press isbn     le roux brigitte and henry rouanet  geometric data analysis from correspondence analysis to structured data analysis dordrecht kluwer a n gorban a y zinovyev principal graphs and manifolds in handbook of research on machine learning applications and trends algorithms methods and techniques olivas e s et al eds information science reference igi global hershey pa usa    wang y klijn j g zhang y sieuwerts a m look m p yang f talantov d timmermans m meijer van gelder m e yu j et al  gene expression profiles to predict distant metastasis of lymph node negative primary breast cancer the lancet     doi   s     data online zinovyev a vidaexpert multidimensional data visualization tool institut curie paris free for non commercial use a n gorban b kegl d c wunsch a zinovyev eds principal manifolds for data visualisation and dimension reduction lncse  springer berlin heidelberg new york  isbn      lu haiping plataniotis k n venetsanopoulos a n  a survey of multilinear subspace learning for tensor data pdf pattern recognition     doi   j patcog    kriegel h p kr√∂ger p schubert e zimek a  a general framework for increasing the robustness of pca based correlation clustering algorithms scientific and statistical database management lecture notes in computer science    isbn      doi         network component analysis reconstruction of regulatory signals in biological systems pdf retrieved february   further reading edit jackson j e  a user s guide to principal components wiley jolliffe i t  principal component analysis springer verlag p  isbn      doi   b jolliffe i t  principal component analysis second edition springer husson fran√ßois l s√©bastien pag√®s j√©r√¥me  exploratory multivariate analysis by example using r chapman hall crc the r series london p isbn     pag√®s j√©r√¥me  multiple factor analysis by example using r chapman hall crc the r series london  pexternal links edit wikimedia commons has media related to principal component analysis university of copenhagen video by rasmus bro on youtubestanford university video by andrew ng on youtubea tutorial on principal component analysisa layman s introduction to principal component analysis on youtube a video of less than  seconds statquest principal component analysis pca clearly explained on youtubesee also the list of software implementationsvtestatisticsoutlineindexdescriptive statisticscontinuous datacentermeanarithmeticgeometricharmonicmedianmodedispersionvariancestandard deviationcoefficient of variationpercentilerangeinterquartile rangeshapemomentsskewnesskurtosisl momentscount dataindex of dispersionsummary tablesgrouped datafrequency distributioncontingency tabledependencepearson product moment correlationrank correlationspearman s rhokendall s taupartial correlationscatter plotgraphicsbar chartbiplotbox plotcontrol chartcorrelogramfan chartforest plothistogrampie chartq q plotrun chartscatter plotstem and leaf displayradar chartdata collectionstudy designpopulationstatisticeffect sizestatistical powersample size determinationmissing datasurvey methodologysamplingstratifiedclusterstandard erroropinion pollquestionnairecontrolled experimentsdesigncontroloptimalcontrolled trialrandomizedrandom assignmentreplicationblockinginteractionfactorial experimentuncontrolled studiesobservational studynatural experimentquasi experimentstatistical inferencestatistical theorypopulationstatisticprobability distributionsampling distributionorder statisticempirical distributiondensity estimationstatistical modellp spaceparameterlocationscaleshapeparametric familylikelihood monotone location scale familyexponential familycompletenesssufficiencystatistical functionalbootstrapuvoptimal decisionloss functionefficiencystatistical distancedivergenceasymptoticsrobustnessfrequentist inferencepoint estimationestimating equationsmaximum likelihoodmethod of momentsm estimatorminimum distanceunbiased estimatorsmean unbiased minimum variancerao blackwellizationlehmann scheff theoremmedian unbiasedplug ininterval estimationconfidence intervalpivotlikelihood intervalprediction intervaltolerance intervalresamplingbootstrapjackknifetesting hypotheses  tailspoweruniformly most powerful testpermutation testrandomization testmultiple comparisonsparametric testslikelihood ratiowaldscorespecific testsz normal student s t testfgoodness of fitchi squaredkolmogorov smirnovanderson darlinglillieforsjarque beranormality shapiro wilk likelihood ratio testmodel selectioncross validationaicbicrank statisticssignsample mediansigned rank wilcoxon hodges lehmann estimatorrank sum mann whitney nonparametric anova way kruskal wallis  way friedman ordered alternative jonckheere terpstra bayesian inferencebayesian probabilitypriorposteriorcredible intervalbayes factorbayesian estimatormaximum posterior estimatorcorrelationregression analysiscorrelationpearson product momentpartial correlationconfounding variablecoefficient of determinationregression analysiserrors and residualsregression model validationmixed effects modelssimultaneous equations modelsmultivariate adaptive regression splines mars linear regressionsimple linear regressionordinary least squaresgeneral linear modelbayesian regressionnon standard predictorsnonlinear regressionnonparametricsemiparametricisotonicrobustheteroscedasticityhomoscedasticitygeneralized linear modelexponential familieslogistic bernoulli binomial poisson regressionspartition of varianceanalysis of variance anova anova analysis of covariancemultivariate anovadegrees of freedomcategorical multivariate time series survival analysiscategoricalcohen s kappacontingency tablegraphical modellog linear modelmcnemar s testmultivariateregressionmanovaprincipal componentscanonical correlationdiscriminant analysiscluster analysisclassificationstructural equation modelfactor analysismultivariate distributionselliptical distributionsnormaltime seriesgeneraldecompositiontrendstationarityseasonal adjustmentexponential smoothingcointegrationstructural breakgranger causalityspecific testsdickey fullerjohansenq statistic ljung box durbin watsonbreusch godfreytime domainautocorrelation acf partial pacf cross correlation xcf arma modelarima model box jenkins autoregressive conditional heteroskedasticity arch vector autoregression var frequency domainspectral density estimationfourier analysiswaveletsurvivalsurvival functionkaplan meier estimator product limit proportional hazards modelsaccelerated failure time aft modelfirst hitting timehazard functionnelson aalen estimatortestlog rank testapplicationsbiostatisticsbioinformaticsclinical trials studiesepidemiologymedical statisticsengineering statisticschemometricsmethods engineeringprobabilistic designprocess quality controlreliabilitysystem identificationsocial statisticsactuarial sciencecensuscrime statisticsdemographyeconometricsnational accountsofficial statisticspopulation statisticspsychometricsspatial statisticscartographyenvironmental statisticsgeographic information systemgeostatisticskrigingcategoryportalcommons wikiproject'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the PCA article does not fall into \n",
    "\n",
    "pca_clean_article = fy.text_cleaner(pca_article)\n",
    "pca_clean_article"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
